tools/c7n_azure/c7n_azure/resources/key_vault.py

```diff
@@ -75,7 +75,11 @@ def __call__(self, i):
         #   - User is whitelisted
         #   - Permissions don't exceed allowed permissions
         for p in i['accessPolicies']:
-            if p[self.key] not in self.users:
+            if self.key not in p:
+                raise KeyError("Key: {0} not found on access policy in Keyvault: {1}. "
```

Comment:
I'm not sure if we should raise an exception here or treat this situation as 'person is not in the white list'.
@stefangordon @aluong do you have any opinion?

Reply:
The trouble here is that the key (principalName for instance) could be missing due to issues in the graph call. Not authorized being one of them, but the graph call could also fail due to transient reasons. I'm a bit concerned that a keyvault that passes the whitelist under normal circumstances could have an action taking against it (like delete) during such an outage, if we were to default to not in the list.

FYI, KeyError was already being thrown for unauthorized cases (principalName was being retrieved), so the only difference here is the test beforehand along with additional information being added.

Reply:
Had a chat with @logachev . If the graph call was indeed successful, then the above behavior recommendation makes sense. Currently considering bubbling up a Boolean to indicate whether or not the graph call was successful and then branch the behavior along the following lines:

- Key missing and not in graph call fields (principalName, aadType, displayName): return false
- Key missing, is graph call field and graph call success: return false
- Key missing, is graph call field and graph call failure: Raise error

Thoughts?

Reply:
So if I understand what @logachev suggests we raise exception on any failed API call but we eliminate failing in scenarios where that call was not required?

This sounds correct.  Open question to me is if we skip the resource or fail the policy.  If it is an intermittent error I'd skip the resource as we normally would, if their principal never had access to the API then this policy will never work as intended and I'd be OK with terminating. 

I'm also open to any more generic way of identifying if the policy needs AD Graph as we probably need similar behavior everywhere it is used, but lets not delay too much trying to find a perfect solution.

Reply:
Modified code to only make API calls when necessary and added a `raise_on_graph_call_error` flag (defaulted to `False` set to true for `WhiteListFiler`'s case) to `GraphHelper` to let the consumer of the class/method choose the behavior they want, since there are other usages. So this should be good now.